### PR TITLE
DCOS-11890: Fix the string to hostname util

### PR DESCRIPTION
**Pull request by orlandohohmeier**
Introduce a new `stringToLabel` util and change the way the `stringToHostname` util works to align them with the respective DC/OS component. The new implementation follows RFC-952 and mirrors the `dcos/mesos_state`  implementation.

Please see dcos/mesos_state#22 as well as dcos/mesos_state#21 for improved/added test cases which illustrate how `dcos/mesos_state` is transform strings.

Closes DCOS-11890

### CODE REVIEW

File: plugins/services/src/js/utils/HostUtil.js [hunk 1]
```diff
@@ -1,32 +1,78 @@
 const HostUtil = {
-  stringToHostname(string) {
-    const HOSTNAME_MAX_LENGTH = 63;
 
+  /**
+   * Util to "transform" a string to a valid hostname following the RFC-952
+   * standard. It mirrors the mirrors the `dcos/mesos_state` implementation.
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: double 'mirrors'

File: plugins/services/src/js/utils/HostUtil.js [hunk 2]
```diff
@@ -1,32 +1,78 @@
...
+   * The util  will remove empty labels and use the string to label util
+   * to ensure that labels don't include invalid characters and don't exceed
+   * the character limit.
+   *
+   * @param {string} string
+   * @return {string} hostename
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: spelling

File: plugins/services/src/js/utils/HostUtil.js [hunk 3]
```diff
@@ -1,32 +1,78 @@
...
+      .join('.');
+  },
 
-    // Strip symbols from beginning
-    while (host.startsWith('-')) {
-      host = host.slice(1);
-    }
+  /**
+   * Util to "convert" a given string to a RFC-952 compliant DNS label.
+   * It mirrors the mirrors the respective `dcos/mesos_state` implementation.
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: mirrors-mirrors

File: plugins/services/src/js/utils/HostUtil.js [hunk 4]
```diff
@@ -1,32 +1,78 @@
...
 
-    // Strip symbols from middle if greater than max length
-    while (host.length > HOSTNAME_MAX_LENGTH && host.charAt(HOSTNAME_MAX_LENGTH - 1) === '-') {
-      host = host.slice(0, HOSTNAME_MAX_LENGTH - 1).concat(host.slice(HOSTNAME_MAX_LENGTH));
+    if (string == null) {
+      return '';
     }
 
-    host = host.slice(0, HOSTNAME_MAX_LENGTH);
+    // Lowercase the input string
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: I think the code is so obvious that don't need the comment.

File: plugins/services/src/js/utils/HostUtil.js [hunk 5]
```diff
@@ -1,32 +1,78 @@
...
+    string = string.toLowerCase();
+
+    // Strip all invalid character or replace them with dashes
+    string = string.replace(/(_)|^-+|-+$|[^a-z0-9-]/g, function (match, p1) {
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: smart 💪

File: plugins/services/src/js/utils/HostUtil.js [hunk 6]
```diff
@@ -1,32 +1,78 @@
...
-      host = host.slice(0, HOSTNAME_MAX_LENGTH - 1).concat(host.slice(HOSTNAME_MAX_LENGTH));
+    if (string == null) {
+      return '';
     }
 
-    host = host.slice(0, HOSTNAME_MAX_LENGTH);
+    // Lowercase the input string
+    string = string.toLowerCase();
+
+    // Strip all invalid character or replace them with dashes
```
Comment (nLight) [inline plugins/services/src/js/utils/HostUtil.js]: let's enhance the comment explicitly mentioning leading/trailing dashes

File: plugins/services/src/js/utils/__tests__/HostUtil-test.js [hunk 7]
```diff
@@ -6,62 +6,162 @@ describe('HostUtil', function () {
...
+      expect(result).toEqual('fdgsf.4abc123');
+    });
+
+    it('should omit dots at the end', function () {
+      const result = HostUtil.stringToHostname('fdgsf.4abc123.');
+
+      expect(result).toEqual('fdgsf.4abc123');
+    });
+
+    it('should omit duplicated dots at the end', function () {
```
Comment (nLight) [inline plugins/services/src/js/utils/__tests__/HostUtil-test.js]: not necessarily at the end ;)

Comment (orlandohohmeier) [inline plugins/services/src/js/utils/__tests__/HostUtil-test.js]: ☑️  Fixed

File: plugins/services/src/js/utils/HostUtil.js (hunk 6 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/utils/HostUtil.js]: ✔️  Done

File: plugins/services/src/js/utils/HostUtil.js (hunk 3 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/utils/HostUtil.js]: ☑️  Fixed

File: plugins/services/src/js/utils/HostUtil.js (hunk 2 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/utils/HostUtil.js]: ☑️  Fixed

File: plugins/services/src/js/utils/HostUtil.js (hunk 1 above)
Comment (orlandohohmeier) [inline plugins/services/src/js/utils/HostUtil.js]: ☑️  Fixed

**Comment by nLight** [comment]
run integration tests

**Comment by orlandohohmeier** [comment]
run integration tests

**Comment by orlandohohmeier** [comment]
run integration tests

**Comment by orlandohohmeier** [comment]
Rebased to include the latest integration test fixes...

**Comment by orlandohohmeier** [comment]
run integration tests

**Comment by jfurrow** [comment]
Nice work!

**Comment by nLight** [comment]
Shall we merge that one?

**Comment by orlandohohmeier** [comment]
🚂  Choo Choo...

**Pull request merged**